added a warning in the doc for get_freer_gpu

fedor-goncharov opened the pull request:
This is a PR related to discussion in 

https://github.com/deepinv/deepinv/issues/563#issuecomment-3005644044


### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

jscanvic reviewed (changes requested):
It looks great so far (especially for a 1st PR!)

I suggest some minor changes & also adding a different fix for users' convenience.

fedor-goncharov commented on deepinv/utils/nn.py:
```diff
@@ -4,6 +4,14 @@
+)
+
 
 def get_freer_gpu(verbose=True):
```
I understand your point, your are right. But if the variable is not set, no need to have an extra parameter `correct`. 
I can just write `if env["CUDA_DEVICE_ORDER"]` - so it will generate an exception and switch to pytorch's routines. 
No warning will be printed. 

However, if the variable is set but to something else we can generate a warning but do nothing more. 
It is a responsibility of the user to choose the right device then.

What do you think?

fedor-goncharov replied:
But have you looked for the latest commit? I kept `os.environ.copy()` so that it is nicer to implement the logic 
on falling back to torch routines. It is following the logic already implemented in the code: if `try-catch` fails somewhere, fallback on torch routines.  

The other way, I am not sure I understand what do you mean on implementation of a `correct` flag. For me 
it means somewhat bypassing `the try-catch` that is already implemented. 

p.s. I did other corrections - put string literal / adjusted docs and text a bit. 

Sorry that a small makes a longer discussion :)

jscanvic replied:
My understanding of what you do is you use `env["CUDA_DEVICE_ORDER"]` which raises an exception if `CUDA_DEVICE_ORDER` is unset to fallback to using `torch.cuda.mem_get_info` in that case. Is that correct?

fedor-goncharov replied:
Hi, is such logic is fine with the project?

jscanvic commented on deepinv/utils/nn.py:
```diff
@@ -29,6 +44,9 @@ def get_freer_gpu(verbose=True):
         idx, mem = np.argmax(memory_available), np.max(memory_available)
         device = torch.device(f"cuda:{idx}")
 
+        env = os.environ.copy()
```
Understood - using `os.getenv` sounds good to me too 👍

fedor-goncharov commented:
I refactored the code for get free gpu. I think larger changes are needed to make the code more clear and consistent:

1. before all logic was implemented only in 1 function only. Now I propose to put 2 different methods in two different functions `_get_freer_gpu_system` for nvidia-smi method and `_get_freer_gpu_torch` for using torch routines. Names of these functions reflect that these are internal methods.  
2. I propose using the PyTorch's method `_get_freer_gpu_torch` as the default one since it is more reliable in any generic situation and is less dependent on tricky parsing of strings from `nvidia-smi`. User can choose which method to choose by setting an appropriate boolean flag. 
3. I added boolean flag to supress all warnings
4. However, if user wants to use `nvidia-smi` method and an exception arises there, then by force the call fallbacks to torch routines and a warning is printed. I think in this situation the warning should never be supressed, as it reports of some problem. User can avoid this message if simply chooses torch's method + supresses all warnings with a flag.
5. The main call is now short and its logic is more clear

Sorry for so significant changes, but in the previous version I could not see how to make it clear with a small change. 
I think it is also not a very good style to implement logic that relies on exceptions, so I decided not to raise exceptions if 
some environment variables are not set.

jscanvic commented:
Thanks for making the changes and sorry for the delay (again!)

Final thing before it can be merged can you please update the changelog and fix the formatting using `black .` to make the checks pass?

fedor-goncharov commented:
I am sorry, I've just noticed, I did not resolve changelog conflicts, could you do it manually? :)

Andrewwango commented:
@jscanvic is it worth adding tests to fix this codecov?
